src/Ztm.WebApi/Watchers/TokenBalance/TokenBalanceWatcher.cs

```diff
@@ -0,0 +1,506 @@
+                    Confirmation = confirmation,
+                    TargetConfirmation = rule.TargetConfirmation,
+                };
+
+                await InvokeCallbackAsync(rule, rule.TimeoutStatus, result, CancellationToken.None);
+            }
+            catch (Exception ex) // lgtm[cs/catch-of-all-exceptions]
+            {
+                this.logger.LogError(ex, "Failed to trigger timeout for address {Address}", address);
+            }
+            finally
+            {
+                this.semaphore.Release();
+            }
+        }
+
+        async Task InvokeCallbackAsync(Rule rule, string status, object data, CancellationToken cancellationToken)
+        {
+            var callback = await this.callbacks.GetAsync(rule.Callback, cancellationToken);
+
+            if (callback == null)
+            {
+                this.logger.LogError("No callback associated with rule {Rule}", rule.Id);
+                return;
+            }
+
+            var result = new CallbackResult(status, data);
+
+            await this.callbacks.AddHistoryAsync(callback.Id, result, cancellationToken);
+            await this.callbackExecutor.ExecuteAsync(callback.Id, callback.Url, result, CancellationToken.None);
+            await this.callbacks.SetCompletedAsyc(callback.Id, CancellationToken.None);
+        }
+
+        async Task IWatcherHandler<Rule, Watch>.AddWatchesAsync(
+            IEnumerable<Watch> watches,
+            CancellationToken cancellationToken)
+        {
+            await this.semaphore.WaitAsync(cancellationToken);
+
+            try
+            {
+                await this.watches.AddAsync(
+                    watches.Where(w => this.watching.ContainsKey(w.Address)),
+                    cancellationToken);
+            }
+            finally
+            {
+                this.semaphore.Release();
+            }
+        }
+
+        async Task<bool> IConfirmationWatcherHandler<Rule, Watch, Confirmation>.ConfirmationUpdateAsync(
+            Confirmation confirm,
+            int count,
+            ConfirmationType type,
+            CancellationToken cancellationToken)
+        {
+            // All watches MUST come from the same rule; otherwise that mean there is a bug somewhere.
+            var rule = confirm.Watches.Select(w => w.Key.Context).Distinct().Single();
+            WatchingInfo watching;
+
+            await this.semaphore.WaitAsync(cancellationToken);
+
+            try
+            {
+                if (!this.watching.ContainsKey(rule.Address))
+                {
+                    // Already timed out.
+                    return false;
+                }
+
+                await this.watches.SetConfirmationCountAsync(confirm.Watches, cancellationToken);
+
+                // Check if completed.
+                if (type == ConfirmationType.Unconfirming || count < rule.TargetConfirmation)
+                {
+                    return false;
+                }
```

Comment:
For this logic we would face the situation that confirmed amount is enough but have more transaction that have confirmations number least than target then when timeout callback would be triggered with timeout data and timeout status. My idea is we should check reached required confirmation amount and if it reach the target we could send callback with confirmed amount and pending amount.

For example If we need 50 XZC and 6 confirmations if we have 40 XZC with 7 confirmations, 10 XZC with 6 confirmations and 1 XZC with 1 confirmation we should send callback as success with message look like
```json
{
  "confirmed amount": 50,
  "pending amount": 1
}
```

Reply:
It will never reached that situation due to if there is a watch that have least confirmation reached target confirmation, which is watch with 10 amount in this case; it will trigger success callback immediately. So, a watch with 1 amount will never happen.

Reply:
After discussion this is possible.